Feat/introduce ADRs (#200)
fishjam-dev/fishjam: docs/adrs/002-change-release-strategy.md

Reviewer: Shouldn't we specify in this document what has to be implemented/modified in the chosen approach?

Author: This is a decision in what direction we want to go, what are the architecture drivers, consequences, other options etc.

 What needs to be done is a topic which should be discussed within the team and should comply with the outcome of this document.

---
status: proposed
date: 17-05-2024
deciders: Kamil Kołodziej, Radosław Szuma, Jakub Pisarek
informed: Fishjam Team, Cloud Team
---

> # Change release strategy to eliminate downtimes

Reviewer 2: **nit**: downtime is uncountable
```suggestion
# Change release strategy to eliminate downtime
```

## Context and Problem Statement

In the current situation when we do a release we are shuting down the server and every room with that.
This is an awful user experience. Another issue is the fact that any ongoing recordings are lost.

## Decision Drivers

> * N o downtime while release happens

Reviewer 2: **nit**: typo
```suggestion
* No downtime while release happens
```

* 0 lost recordings during release

## Considered Options

> * Rolling update with eviction

Reviewer 3: Should we write more on the API? E.g. whether we want to add a new REST endpoint?

* Active room migration

## Decision Outcome

Chosen option: "Rolling update with eviction", because
while second option sounds great, we acknowledge that solution is not trivial and we want to fix downtimes ASAP. First solution is a good starting point to remove existing problems and allow us to move forward.

### Consequences

> Every new release will need to trigger fishjam process which will handle the shutdown. What that effectively means is that we are leaving the responsibility of triggering release to external orchestration tool (what is usually the case) and to monitor if the application exited correctly.

Reviewer 2: **issue**: So, release != deployment or upgrade. I suggest generally replacing this word everywhere in this ADR with something else. IMHO it makes it harder to understand, as this has not much in common with the Fishjam release cycle.

```suggestion
Every new deployment will need to trigger fishjam process which will handle the shutdown. What that effectively means is that we are leaving the responsibility of triggering release to external orchestration tool (what is usually the case) and to monitor if the application exited correctly.
```

Reviewer 4: Seconding this, in favour of 'deployment' as it can refer to both an 'upgrade' and a 'downgrade'/'downpatch'

> External process will trigger fishjam shutdown process which is gonna mark application as one that no longer accepts the connections (new rooms etc.). Once every room is closed and all of the recordings are computed we are going to mark fishjam as ready to shutdown. This application instance should also be effectively removed from any external load balancing (k8s etc.) while it should still allow erlang clustering. Orchestrator process should query the fishjam till it's ready to shutdown and sends sigterm signal.

Reviewer 4: I wouldn't mention 'no longer accepting the connections', this can paint the wrong picture and isn't entirely in line with what we decided (the fishjam that's shutting down should still be able to relay any and all requests to others in the cluster).

This is applicable one by one for every instance in cluster, although it may take some time to release a new version and we may have 2 different versions on the cluster at the same time, we are accepting that tradeoff. We may consider `force` option to release a version no matter the state of fishjam (which may result in downtime).

## Pros and Cons of the Options

### Rolling update with eviction

1. External orchestrator process triggers new release

> 2. Starts new instance of fishjam with new version

Author: Then we will have (N-1) instances in cluster available to handle the same amount of traffic.

3. Fishjam shutdown process start

> 4. Stop receiving traffic to that instance

Reviewer: Shouldn't we specify which traffic we won't receive? If I remember correctly, we mentioned only requests about creating a new room; all other traffic should be allowed.

> 4. Stop receiving traffic to that instance

Reviewer 2: **issue**: It should be specified **who** is responsible for directing the traffic to the right instance, especially with future Cluster API changes.

Reviewer 4: Right now, business logic (BE); after the API changes, every FJ in the cluster will be able to route the requests.

I don't think this is the right place to specify that, I think a separate ADR about the new clustering behaviour should (and will) be created; then we can simply link it here. Alternatively, we could add a single sentence like the first one in this comment

Reviewer 2: That would be good enough. I'm simply wondering whether it is supposed to be implemented as a load-balancer configuration or will it be done automagically by Fishjam.

5. Wait till the last room is closed and all recordings are completed

> 6. Finish fishjam shutdown process - ready for sigterm

Reviewer 2: **strong suggestion**: In most systems/contexts (docker, k8s, bare-metal) `SIGTERM` should be the initial signal sent to the application when requesting a graceful shutdown. I strongly suggest outlining this fact here.

I think we can easily call this a contract between the Cloud and Fishjam teams.

Reviewer 5: note that grace period between sigterm and sigkill is brief for most orchestrators iirc like 30 seconds by default) and we may want to handle it properly

Reviewer 2: All considered platforms support an infinite (or long enough) grace period for it not to be a problem. The user or the cloud team will just have to handle it on their side.

> 7. External orchestrator process kills fishjam instance

Reviewer 2: **issue**: When the instance receives a `SIGTERM` - after cleaning up, it should quit by itself, not on an additional orchestrator request.

> 8. Repeat for rest of the remaining instances

Reviewer 2: **suggestion**: This step depends on the orchestrator's configuration and is more of a suggestion than a strict requirement. I'd focus on signal handling and clustering behavior in this ADR, not the exact external behavior.

* Good, because we will eliminate downtimes with releases
* Good, because we won't lose any recordings

> * Neutral, because we require some external work from the orchestrator

Reviewer 6: This is "bad", not "neutral" and should be removed by considering the following (some things were already mentioned in the other comments):
* The ADR should cover how to pass the info about shutting down. Remember that REST endpoint would have to be somehow secured to be available only to some orchestrator, so using SIGTERM seems like an easier option
* FJ should not expect orchestrator to check shutdown readiness after sending "soft-shutdown" request and should shutdown automatically

Author: already resolved in [74ec204](https://github.com/fishjam-dev/fishjam/pull/200/commits/74ec20407a0aded6e052828b3fd5d1d0c4f1a898)

* Bad, because release process may take some time (effectively as long as the longest conversation/stream)
* Bad, because we may end up with different versions on cluster at the same time

###  Active room migration

This solution wasn't researched much so we supposed the flow should be like that:

1. External orchestrator process triggers new release
2. Starts new instance of fishjam with new version
3. Fishjam starts migrating rooms to new instance
4. Somehow handles the recordings (?)
5. Once peers/rooms/streams are migrated app is ready for sigterm
6. External orchestrator process kills fishjam instance
7. Repeat for rest of the remaining instances

* Good, because we will eliminate downtimes with releases
* Good, because we won't lose any recordings
* Good, because it happens fast, we don't have to wait for rooms to close/streams to end

> * Good, because we will have 2 different versions on cluster for a small amount of time

Reviewer 2: **nit**: grammar
```suggestion
* Good, because we will have 2 different versions on cluster for a short amount of time
```

* Neutral, because we require some external work from the orchestrator

> * Bad, because we don't have a clue how to handle the active peer migration to new instance right now

Author: Haha! Yeah that's right I will correct this.

> * Bad, because we don't have a clue how to handle the active peer migration to new instance right now

Reviewer 2: I don't agree. ;P
The cost of a solution is definitely a drawback that is considered when making a decision. Because ADR is quite literally, a record of a decision, it should contain cost doubts and rationale for why we've made a specific choice at this time.

* Bad, because we don't know how to handle the meetings with recording enabled during the migration

## More Information

> This is decision which is heavily dependent on the Cloud Team and may be change soon to meet their requirements.

Reviewer 2: **nit**: grammar
```suggestion
This decision is heavily dependent on the Cloud Team and may be changed soon to meet their requirements.
```
